src/apps/api/views/competitions.py

```diff
@@ -211,59 +210,107 @@ def email_all_participants(self, request, pk):
 
         # Query Needed data and filter to what is needed.
         phase_pks = [phase.id for phase in Phase.objects.filter(competition_id=competition.id)]
-        submission_query = Submission.objects.filter(Q(phase_id__in=phase_pks) & Q(has_children=False) & Q(leaderboard_id__isnull=False))
+        submission_query = Submission.objects.filter(
+            Q(phase_id__in=phase_pks) & Q(has_children=False) & Q(leaderboard_id__isnull=False))
         if not submission_query.exists():
             raise ValidationError("There are no submissions on the leaderboard")
 
-        # Build the data needed for the csv's into a dictionary
-        csv = {}
-        user_id = set()
-        for sub in submission_query:
-            if sub.leaderboard_id not in csv.keys():
-                csv[sub.leaderboard_id] = {}
-                csv[sub.leaderboard_id]["user"] = ["Username"]
-                columns = Column.objects.filter(leaderboard_id=sub.leaderboard_id)
+        # Build the data needed for the leaderboard_data's into a dictionary
+        leaderboard_data = {}
+        leaderboardTitles = {}
+        columnsTitles = {}
+        for submission in submission_query:
+            leaderID = submission.leaderboard_id
+            if leaderID not in leaderboardTitles.keys():
+                leaderboard = Leaderboard.objects.prefetch_related('columns').get(pk=leaderID)
+                columnsTitles.update({leaderID: {col.id: f'{col.title}({col.id})' for col in leaderboard.columns.all()}})
+                leaderboardTitles.update({leaderID: f"{leaderboard.title}({leaderID})"})
+                leaderboard_data[leaderboardTitles[leaderID]] = {}
+            leaderboard_data[leaderboardTitles[leaderID]][submission.owner.username] = {}
+            for score in submission.scores.all():
+                leaderboard_data[leaderboardTitles[leaderID]][submission.owner.username].update({columnsTitles[leaderID][score.column_id]: float(score.score)})
+        return leaderboard_data
+
+    @action(detail=True, methods=['GET'], renderer_classes=[JSONRenderer, CSVRenderer, ZipRenderer])
+    def results(self, request, pk, format=None):
+        competition = self.get_object()
+        self.collect_leaderboard_check_permissions(request, competition)
+        data = self.collect_leaderboard_data(competition)
+        selected_data = {}
+        selected_id = request.GET.get('id')
+        selected_leaderboard = request.GET.get('title')
+
+        if format == 'zip':
+            with SpooledTemporaryFile() as tmp:
+                with zipfile.ZipFile(tmp, 'w', zipfile.ZIP_DEFLATED) as archive:
+                    for leaderboard in data:
+                        temp_file = NamedTemporaryFile()
+                        line = "Username"
+                        columns = data[leaderboard][list(data[leaderboard].keys())[0]]
+                        for column_title in columns:
+                            line += f',{column_title}'
```

Comment:
we should be using a `CSVDictReader` + `CSVDictWriter`, ideally

https://docs.python.org/3/library/csv.html